src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -312,6 +313,66 @@ impl ScalarFunc {
             _ => Err(box_err!("invalid direction value: {}", direction)),
         }
     }
+
+    #[inline]
+    pub fn substring_index<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let delim = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
+        let count = try_opt!(self.children[2].eval_int(ctx, row));
+        if delim.is_empty() || count == 0 {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        let r = if mysql::has_unsigned_flag(self.children[2].get_tp().get_flag()) {
+            substring_index_positive(&s, delim.as_ref(), count as u64 as usize)
+        } else if count >= 0 {
+            substring_index_positive(&s, delim.as_ref(), count as usize)
+        } else {
+            let count = if count == i64::min_value() {
+                i64::max_value() as usize + 1
+            } else {
+                -count as usize
+            };
+            substring_index_negative(&s, delim.as_ref(), count)
+        };
+        Ok(Some(Cow::Owned(r.into_bytes())))
+    }
+}
+
+#[inline]
+fn substring_index_positive(s: &str, delim: &str, count: usize) -> String {
+    let mut bg = 0;
+    let mut cnt = 0;
+    let mut last = 0;
+    while cnt < count {
+        if let Some(idx) = s[bg..].find(delim) {
+            last = bg + idx;
+            bg = last + delim.len();
+            cnt += 1;
+        } else {
+            last = s.len();
+            break;
+        }
+    }
+    s[..last].to_string()
+}
+
+#[inline]
+fn substring_index_negative(s: &str, delim: &str, count: usize) -> String {
+    let mut bg = 0;
+    let mut positions = VecDeque::with_capacity(count.min(128));
+    positions.push_back(0);
+    while let Some(idx) = s[bg..].find(delim) {
+        bg = bg + idx + delim.len();
+        positions.push_back(bg);
```

Comment:
If you choose to do `push_back` first and the length of `positions` is `count`. It may trigger the memory reallocation.